desc/basis.py

```diff
@@ -1413,6 +1413,424 @@ def zernike_radial_poly(r, l, m, dr=0, exact="auto"):
     return polyval_vec(coeffs, r, prec=prec).T
 
 
+def zernike_radial_direct(r, n, m):
+    """Radial part of zernike polynomials. Modified Prata's algorithm."""
+    R = np.zeros((n, m))  # R_nmk
+
+    M = 10  # dummy number for now
+    for n in range(M):
+        K_1 = 2 * n / (m + n)
+        K_2 = 1 - K_1
+
+        R[n, n] = r**n
+
+        for m in range(n - 2, 1, -2):
+            R[n, m] = r * K_1 * R[n - 1, m - 1] + K_2 * R[n - 2, m]
+
+            if n % 2 == 0:
+                R[n, 0] = 2 * r * R[n - 1, 1] - R[n - 2, 0]
+
+            R_nm = np.sum([R * r**k for k in range(M)])
+
+    # not sure if this part is needed
+    R_k = np.zeros((n - m) / 2)
+    for k in range((n - m) / 2):
+        R_k[k] += (
+            (-1) ** ((n - k) / 2)
+            * factorial((n + k) / 2)
+            / factorial((n - k) / 2)
+            * factorial((k + abs(m)) / 2)
+            * factorial((k - abs(m)) / 2)
+        )
+
+    return R_nm
+
+
+@functools.partial(jit, static_argnums=3)
+def zernike_radial_optimized(r, l, m, dr=0):  # noqa: C901
+    """Radial part of zernike polynomials.
+
+    This is completely JITable version of zernike_radial_optimized
+    function.
+
+    Parameters
+    ----------
+    r : ndarray, shape(N,)
+        radial coordinates to evaluate basis
+    l : ndarray of int, shape(K,)
+        radial mode number(s)
+    m : ndarray of int, shape(K,)
+        azimuthal mode number(s)
+
+    Returns
+    -------
+    out : ndarray, shape(N,K)
+        basis function(s) evaluated at specified points
+
+    """
+    if dr > 4:
+        raise NotImplementedError(
+            "Analytic radial derivatives of Zernike polynomials for order>4 "
+            + "have not been implemented."
+        )
+
+    def body_inner(N, args):
+        r_jacobi, alpha, r, out, P_past, m, n = args
+        P_n1, P_n2 = P_past
+        # Calculate Jacobi polynomial for m,n pair
+        P_n = jacobi_poly_single(r_jacobi, N, alpha, 0, P_n1, P_n2)
+
+        # Find the index corresponding to the original array
+        index = jnp.where(
+            jnp.logical_and(m == alpha, n == N),
+            jnp.arange(m.size),
+            0,
+        )
+        # jnp.where() will result in 1D array with 1 value which
+        # is not 0. Sum it to find the index
+        idx = jnp.sum(index)
```

Comment:
Are we sure there will only be one nonzero value? Might need to also test for stuff with duplicate modes etc

Reply:
Yeah I assumed there are no duplicate modes. And also I assumed that mode exists. Since I am constructing in between modes from the max M, if for example for Mmax=10, there is actually no m=5th mode, this code will change the 0th index of the out array. It's a good idea to some checks before that

Reply:
Actually, these assumptions caused problems for existing tests with specified modes
l = np.array([3, 4, 6])
m = np.array([1, 2, 2])
I will look for it in detail.

Reply:
I think this indexing along with the assignment below could be replaced by

```python
out = jnp.where(jnp.logical_and(m == alpha, n == N), (-1) ** N * r**alpha * P_n, out)
```

Which I think should also work for duplicate modes or non-sequential stuff

Reply:
I believe I tried something like this before (and tried again now). The problem is out is the full result array, so it is 2D with (r.size, m.size). However, jnp.logical_and(m == alpha, n == N) is 1D with m.size and (-1) ** N * r**alpha * P_n is 1D with r.size . Dimensions cause broadcasting issue.

Reply:
Can't you just broadcast things like normal?

```python
mask = jnp.logical_and(m == alpha, n == N)
val = (-1) ** N * r**alpha * P_n
out = jnp.where(mask[None,:], val[:,None], out)
```

Reply:
In practice there will often be duplicate modes. The most common reason is if N (toroidal resolution) is nonzero, in which case there will be a L,M pair for each N, but the zernike radial part will see that as duplicates for L,M